go-controller/pkg/node/node.go

```diff
@@ -250,10 +252,45 @@ func (n *OvnNode) Start() error {
 	if !ok {
 		return fmt.Errorf("Cannot get kubeclient for starting CNI server")
 	}
+	err = n.WatchEndpoints()
+	if err != nil {
+		return err
+	}
 
 	// start the cni server
 	cniServer := cni.NewCNIServer("", kclient.KClient)
 	err = cniServer.Start(cni.HandleCNIRequest)
 
 	return err
 }
+
+func (n *OvnNode) WatchEndpoints() error {
+	_, err := n.watchFactory.AddEndpointsHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(old, new interface{}) {
+			epNew := new.(*kapi.Endpoints)
+			epOld := old.(*kapi.Endpoints)
+			if reflect.DeepEqual(epNew.Subsets, epOld.Subsets) {
+				return
+			}
+			newEndpointAddressMap := buildEndpointAddressMap(epNew.Subsets)
+			for _, subset := range epOld.Subsets {
+				for _, address := range subset.Addresses {
+					if !newEndpointAddressMap[address.IP] {
+						deleteConntrack(address.IP)
+					}
+				}
+			}
+		},
+	}, nil)
+	return err
+}
+
+func buildEndpointAddressMap(epSubsets []kapi.EndpointSubset) map[string]bool {
+	addressMap := make(map[string]bool)
```

Comment:
nit you could use an empty struct instead of a bool and check if the key exists in your other function to save a tiny bit of memory

Reply:
done